controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -83,41 +86,68 @@
-                                                             // 2. seal latest collection
-                                                             .thenCompose(v -> storeHelper.createZNodeIfNotExist(
-                                                                     getCollectionSealedPath(scope, stream, latestcollectionNum)))
-                                                             // 3. call addEntity recursively
-                                                             .thenCompose(v -> addEntity(scope, stream, entity))
-                                                             // 4. delete empty sealed collection path
-                                                             .thenCompose(orderedPosition -> 
-                                                                     tryDeleteSealedCollection(scope, stream, latestcollectionNum)
-                                                                     .thenApply(v -> orderedPosition));
-
-                                       } else {
-                                           return CompletableFuture.completedFuture(Position.toLong(latestcollectionNum, position));
-                                       }
-                                   }))
+                                .thenCompose(positionPath -> {
+                                    int position = getPositionFromPath(positionPath);
+                                    if (position > rollOverAfter) {
+                                        // if newly created position exceeds rollover limit, we need to delete that entry
+                                        // and roll over.
+                                        try {
+                                            // 1. delete newly created path
+                                            CuratorOp op1 = storeHelper.getClient().transactionOp().delete().forPath(positionPath);
+
+                                            // 2. seal latest collection
+                                            String sealLatestCollectionPath = getCollectionSealedPath(scope, stream, latestCollectionNumber);
+                                            CuratorOp op2 = storeHelper.getClient().transactionOp().create().forPath(sealLatestCollectionPath);
+
+                                            // 3. create new collection and put the entity in
+                                            int newCollectionNumber = latestCollectionNumber + 1;
+
+                                            String collectionPath = getCollectionPath(scope, stream, newCollectionNumber);
+                                            String entitiesPath = getEntitiesPath(scope, stream, newCollectionNumber);
+                                            String entitySequentialPath = getEntitySequentialPath(scope, stream, newCollectionNumber);
+
+                                            CuratorOp op31 = storeHelper.getClient().transactionOp().create().forPath(collectionPath);
```

Comment:
When using multi-op(a.k.a zk's transaction, but I'd like to distinguish with zk's txn which is a ZK internal concept), we can only commit so-called primitive zk ops. Thus, we lost some power from curator includes `createParentIfNeeded` and so on. It might make the implement a bit clumsy. I'm open to your idea whether we add more utils on multi-ops(in ZKStoreHelper, maybe).